pallet/src/lib.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Encode;
+
+use beefy_primitives::{AuthorityIndex, ConsensusLog, BEEFY_ENGINE_ID};
+use frame_support::{decl_module, decl_storage, Parameter};
+use sp_runtime::{
+	generic::DigestItem,
+	traits::{IsMember, Member},
+	RuntimeAppPublic,
+};
+use sp_std::prelude::*;
+
+pub trait Trait: frame_system::Trait {
+	/// The identifier type for an authority.
+	type AuthorityId: Member + Parameter + RuntimeAppPublic + Default;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Beefy {
+		/// The current authorities
+		pub Authorities get(fn authorities): Vec<T::AuthorityId>;
+	}
+	add_extra_genesis {
+		config(authorities): Vec<T::AuthorityId>;
+		build(|config| Module::<T>::initialize_authorities(&config.authorities))
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin { }
+}
+
+impl<T: Trait> Module<T> {
+	fn change_authorities(new: Vec<T::AuthorityId>) {
+		<Authorities<T>>::put(&new);
+
+		let log: DigestItem<T::Hash> =
+			DigestItem::Consensus(BEEFY_ENGINE_ID, ConsensusLog::AuthoritiesChange(new).encode());
+
+		<frame_system::Module<T>>::deposit_log(log);
+	}
+
+	fn initialize_authorities(authorities: &[T::AuthorityId]) {
+		if !authorities.is_empty() {
+			assert!(
+				<Authorities<T>>::get().is_empty(),
+				"Authorities are already initialized!"
+			);
+			<Authorities<T>>::put(authorities);
+		}
+	}
+}
+
+impl<T: Trait> sp_runtime::BoundToRuntimeAppPublic for Module<T> {
+	type Public = T::AuthorityId;
+}
+
+impl<T: Trait> pallet_session::OneSessionHandler<T::AccountId> for Module<T> {
+	type Key = T::AuthorityId;
+
+	fn on_genesis_session<'a, I: 'a>(validators: I)
+	where
+		I: Iterator<Item = (&'a T::AccountId, T::AuthorityId)>,
+	{
+		let authorities = validators.map(|(_, k)| k).collect::<Vec<_>>();
+		Self::initialize_authorities(&authorities);
```

Comment:
Wouldn't that panic in case we have authorities in the chain spec already?

Reply:
We guard against this inside `initialize_authorities` and only set the authorities if they're not set already. In the case where we have authorities defined in the module genesis config and also on genesis session I don't know which would take precedence, I think it would depend on module order when assembling runtime.

Reply:
I took this code from the aura pallet and misread it now when you commented, so my comment above doesn't make much sense. The code will indeed panic if there are authorities defined both on the module genesis config and as part of genesis session. This is to make sure the end-user doesn't try to initialize a chain using both mechanisms.